REMEMBER: FIRST CREATE A COPY OF THIS FILE WITH A UNIQUE NAME AND DO YOUR WORK THERE. AND MAKE SURE YOU COMMIT YOUR CHANGES TO THE `hw3_submissions` BRANCH.

# Assignment 3 | Cleaning and Exploring Data with Pandas



<img src="data/scoreCard.jpg" width=250>

In this assignment, you will investigate restaurant food safety scores for restaurants in San Francisco. Above is a sample score card for a restaurant. The scores and violation information have been made available by the San Francisco Department of Public Health. 

## Loading Food Safety Data


There are 2 files in the data directory:
1. business.csv containing food establishments in San Francisco
1. inspections.csv containing retaurant inspections records

Let's start by loading them into Pandas dataframes.  One of the files, business.csv, has encoding (ISO-8859-1), so you will need to account for that when reading it.

### Question 1

#### Question 1a
Read the two files noted above into two pandas dataframes named `bus` and `ins`, respectively. Print the first 5 rows of each to inspect them.


In [222]:
import pandas as pd

In [223]:
bus = pd.read_csv('businesses.csv', encoding='ISO-8859-1')
ins = pd.read_csv('inspections.csv')

In [224]:
bus.head()

business_id                                name  \
0           10                    TIRAMISU KITCHEN   
1           19               NRGIZE LIFESTYLE CAFE   
2           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3           31      NORMAN'S ICE CREAM AND FREEZES   
4           45                 CHARLIE'S DELI CAFE   

                         address           city state postal_code   latitude  \
0                  033 BELDEN PL  San Francisco    CA       94104  37.791116   
1   1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
2  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
3           2801 LEAVENWORTH ST   San Francisco    CA       94133  37.807155   
4                3202 FOLSOM ST   San Francisco    CA       94110  37.747114   

    longitude  phone_number  
0 -122.403816  +14154217044  
1 -122.421547  +14157763262  
2 -122.403135  +14156779494  
3 -122.419004           NaN  
4 -122.413641  +14156415051

In [225]:
ins.head()

business_id  score      date     type
0           10     82  20160503  routine
1           10     94  20140729  routine
2           10     92  20140114  routine
3           19     94  20160513  routine
4           19     94  20141110  routine

## Examining the Business data

From its name alone, we expect the `businesses.csv` file to contain information about the restaurants. Let's investigate this dataset.

### Question 2

#### Question 2a: How many records are there?

In [226]:
print("There are:", bus.shape[0], "records in business csv")

There are: 6315 records in business csv


#### Question 2b: How many unique business IDs are there?  

In [227]:
print("There are:", bus['business_id'].nunique(), "unique business IDs in business csv")  # len(bus['business_id'].unique())

There are: 6315 unique business IDs in business csv


#### Question 2c: What are the 5 most common businesses by name, and how many are there in San Francisco?

In [228]:
bus_sf = bus[bus['city'] == 'San Francisco']
top_5_names = bus_sf['name'].value_counts().head(5) 
print("Top 5 most common businesses by name and thier corresponding quantities in San Francisco:", top_5_names)

Top 5 most common businesses by name and thier corresponding quantities in San Francisco: name
STARBUCKS COFFEE              72
PEET'S COFFEE & TEA           24
MCDONALDS                     12
SAN FRANCISCO SOUP COMPANY    11
WALGREENS                     11
Name: count, dtype: int64


## Zip code

Next, let's  explore some of the variables in the business table. We begin by examining the postal code.

### Question 3

#### Question 3a
How are the zip code values stored in python (i.e. data type)?

To answer this you might want to examine a particular entry.

In [229]:
type(bus['postal_code'][0])

str

In [230]:
print("In Python, based on the provided output, the zip code values from the 'postal_code' column in bus dataframe are stored as strings (data type: 'str' or 'object')")
print("Zip codes are typically composed of numerical digits, but they are treated as identifiers rather than actual numeric values for mathematical operations")
print("Storing zip codes data as strings preserves any leading zeros and ensures they are not inadvertently used in arithmetic operations.")

In Python, based on the provided output, the zip code values from the 'postal_code' column in bus dataframe are stored as strings (data type: 'str' or 'object')
Zip codes are typically composed of numerical digits, but they are treated as identifiers rather than actual numeric values for mathematical operations
Storing zip codes data as strings preserves any leading zeros and ensures they are not inadvertently used in arithmetic operations.


#### Question 3b

What are the unique values of postal_code?

In [231]:
bus['postal_code'].unique()

array(['94104', '94109', '94133', '94110', '94122', '94103', '94115',
       '94131', '94111', '94117', '94107', '94108', '94102', '94132',
       '94105', '94134', '94124', '94116', '94121', '94118', '94112',
       '94123', '94114', '94127', '941102019', '941', '94545', '94066',
       '941033148', nan, '94158', '95105', '94013', '94130', 'CA',
       '92672', '94120', '94143', '94609', '94101', '00000', '94188',
       '94621', '94014', '94129', '94602', 'Ca'], dtype=object)

#### Question 3c

Let's say we decide to exclude the businesses that have no zipcode for our analysis (which might include food trucks for example).  Use the list of valid 5-digit zip codes below to create a new dataframe called bus_valid, with only businesses whose postal_codes show up in this list of valid zipcodes. How many businesses are there in this new dataframe?

In [232]:
nan_values = bus['postal_code'].isna().sum()  # isnull
print("There are", len(bus['postal_code']), "zip code values stored in total")
print("But there are", nan_values, "NaN value")
print("Thus, there are", (len(bus['postal_code']) - nan_values), "zip code values acutally stored")

There are 6315 zip code values stored in total
But there are 259 NaN value
Thus, there are 6056 zip code values acutally stored


In [233]:
validZip = ["94102", "94103", "94104", "94105", "94107", "94108",
            "94109", "94110", "94111", "94112", "94114", "94115",
            "94116", "94117", "94118", "94121", "94122", "94123", 
            "94124", "94127", "94131", "94132", "94133", "94134"]

In [234]:
vb = bus[bus['postal_code'].isin(validZip)]
print("There are", len(vb['name']), "businesses in total are in the new dataframe")
print("There are", vb['name'].nunique(), "unique businesses are in the new dataframe")

There are 5999 businesses in total are in the new dataframe
There are 5438 unique businesses are in the new dataframe


## Latitude and Longitude

Another aspect of the data we want to consider is the prevalence of missing values. If many records have missing values then we might be concerned about whether the nonmissing values are representative of the population.

### Question 4
 
Consider the longitude and latitude in the business DataFrame. 

#### Question 4a

How many businesses are missing longitude values, working with only the businesses that are in the list of valid zipcodes?

In [235]:
print("There are", len(vb[pd.isnull(vb['longitude'])]), 'values are missing longitude values in the new business dataframe')

There are 2483 values are missing longitude values in the new business dataframe


#### Question 4b

Create a new dataframe with one row for each valid zipcode. The dataframe should include the following three columns:

1. `postal_code`: Contains the zip codes in the `validZip` variable above.
2. `null_lon`: The number of businesses in that zipcode with missing `longitude` values.
3. `not_null_lon`: The number of businesses without missing `longitude` values.

In [236]:
result_rows = []

for zip_code in validZip:
    df_filtered = vb[vb['postal_code'] == zip_code]
    null_lon_count = df_filtered['longitude'].isna().sum()
    not_null_lon_count = df_filtered['longitude'].notna().sum()
    result_rows.append({'postal_code': zip_code, 'null_lon': null_lon_count, 'not_null_lon': not_null_lon_count})      # Append the counts to the result list

result_df = pd.DataFrame(result_rows)  # Convert the list of results into a DataFrame
result_df

postal_code  null_lon  not_null_lon
0        94102       196           262
1        94103       252           306
2        94104        59            74
3        94105       100           126
4        94107       247           204
5        94108        82           142
6        94109       146           237
7        94110       251           334
8        94111       115           167
9        94112        59           131
10       94114       108           122
11       94115        76           162
12       94116        35            58
13       94117        67           130
14       94118       100           143
15       94121        55           106
16       94122       123           147
17       94123        53           116
18       94124        88            76
19       94127        28            43
20       94131        14            34
21       94132        65            69
22       94133       135           282
23       94134        29            45

#### 4c. Do any zip codes appear to have more than their 'fair share' of missing longitude? 

To answer this, you will want to compute the proportion of missing longitude values for each zip code, and print the proportion missing longitude, and print the top five zipcodes in descending order of proportion missing postal_code.

In [237]:
result_rows = []

for zip_code in validZip:
    df_filtered = vb[vb['postal_code'] == zip_code]
    null_lon_count = df_filtered['longitude'].isna().sum()
    not_null_lon_count = df_filtered['longitude'].notna().sum()
    total_rows_count = len(df_filtered)
    proportion_missing = null_lon_count / total_rows_count if total_rows_count > 0 else 0
    result_rows.append({
        'postal_code': zip_code, 
        'null_lon': null_lon_count, 
        'not_null_lon': not_null_lon_count,
        'proportion_missing_lon': proportion_missing
    })

result_df = pd.DataFrame(result_rows)
top_missing_lon_df = result_df.sort_values(by='proportion_missing_lon', ascending=False)[['postal_code', 'proportion_missing_lon']]
top_missing_lon_df.head(5)

postal_code  proportion_missing_lon
4        94107                0.547672
18       94124                0.536585
21       94132                0.485075
10       94114                0.469565
16       94122                0.455556

In [238]:
print("Zip codes 94107, 94124, 94132, 94114, and 94122 appear to have more than their 'fair share' of missing longitude since their proportion of missing longitude is over 45%")

Zip codes 94107, 94124, 94132, 94114, and 94122 appear to have more than their 'fair share' of missing longitude since their proportion of missing longitude is over 45%


# Investigate the inspection data

Let's now turn to the inspection DataFrame. Earlier, we found that `ins` has 4 columns, these are named `business_id`, `score`, `date` and `type`.  In this section, we determine the granularity of `ins` and investigate the kinds of information provided for the inspections. 

### Question 5

#### Question 5a
As with the business data, assess whether there is one inspection record for each business, by counting how many rows are in the data and how many unique businesses there are in the data. If they are exactly the same number, it means there is only one inspection per business, clearly.

In [239]:
total_rows = len(ins)
unique_businesses = ins['business_id'].nunique()
print("There are", total_rows, "records in total, and there are",  unique_businesses, "unique unique business records")

There are 15430 records in total, and there are 5730 unique unique business records


#### Question 5b

What values does `type` take on? How many occurrences of each value is in the DataFrame? Create a new dataframe named `ins2` by copying `ins` and keeping only records with values of `type` that occur more than 10 times in the original table. In other words, eliminate records that have values of `type` that occur rarely (< 10 times). Check the result to make sure rare types are eliminated.

In [240]:
ins['type'].unique()

array(['routine', 'complaint'], dtype=object)

In [241]:
type_counts = ins['type'].value_counts()
types_more_than_10 = type_counts[type_counts > 10].index

In [242]:
print("type column is string, it has two different value: 'routine' and 'complaint'.")
print("\nThis is the occurrences of each two value:")
print(type_counts)
print("\nSo value 'routine occurs more than 10 times'")

type column is string, it has two different value: 'routine' and 'complaint'.

This is the occurrences of each two value:
type
routine      15429
complaint        1
Name: count, dtype: int64

So value 'routine occurs more than 10 times'


In [243]:
ins2 = ins[ins['type'].isin(types_more_than_10)]
type_counts_new = ins2['type'].value_counts()
print("This is the occurrences of ins2 type column:")
print(type_counts_new)

This is the occurrences of ins2 type column:
type
routine    15429
Name: count, dtype: int64


#### Question 5c

Since the data was stored in a .csv file, the dates are formatted as strings such as `20160503`. Once we read in the data, we would like to have dates in an appropriate format for analysis. Add a new column called `year` by capturing the first four characters of the date column. 

Hint: we have seen multiple ways of doing this in class, includings `str` operations, `lambda` functions, `datetime` operations, and others. Choose the method that works best for you :)

In [244]:
'''
ins2['date'] = ins2['date'].astype(str)
ins2['year'] = ins2['date'].str[:4]  # capture the first four characters of the 'date' string
ins2.head()
'''

ins2['date'] = pd.to_datetime(ins2['date'], format='%Y%m%d')
ins2['year'] = ins2['date'].dt.year
ins2.head() 

C:\Users\Cedric\AppData\Local\Temp\ipykernel_25608\817708507.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  ins2['date'] = pd.to_datetime(ins2['date'], format='%Y%m%d')
C:\Users\Cedric\AppData\Local\Temp\ipykernel_25608\817708507.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  ins2['year'] = ins2['date'].dt.year


business_id  score       date     type  year
0           10     82 2016-05-03  routine  2016
1           10     94 2014-07-29  routine  2014
2           10     92 2014-01-14  routine  2014
3           19     94 2016-05-13  routine  2016
4           19     94 2014-11-10  routine  2014

#### Question 5d

What range of years is covered in this data set? Are there roughly same number of inspections each year? Try dropping records for any years with less than 50 inspections and store the result in a new dataframe named `ins3`.

In [245]:
year_range = ins2['year'].agg(['min', 'max'])  # the range of years
print("The year range covered in 'ins2' data set:")
print(year_range)

The year range covered in 'ins2' data set:
min    2013
max    2016
Name: year, dtype: int32


In [246]:
# Counting the number of inspections per year
inspections_per_year = ins2['year'].value_counts()
print(inspections_per_year)
print('\nAccording to the number of inspections per year, I think year 2016 and year 2014 share roughly the same number of inspections, but year 2015 and year 2013 are pretty less')

year
2016    5839
2014    5629
2015    3923
2013      38
Name: count, dtype: int64

According to the number of inspections per year, I think year 2016 and year 2014 share roughly the same number of inspections, but year 2015 and year 2013 are pretty less


In [247]:
years_with_50_or_more_inspections = inspections_per_year[inspections_per_year >= 50].index
ins3 = ins2[ins2['year'].isin(years_with_50_or_more_inspections)]

inspections_per_year_filtered = ins3['year'].value_counts()
inspections_per_year_filtered

year
2016    5839
2014    5629
2015    3923
Name: count, dtype: int64

Let's examine only the inspections for one year: 2016. This puts businesses on a more equal footing because [inspection guidelines](https://www.sfdph.org/dph/eh/Food/Inspections.asp) generally refer to how many inspections should occur in a given year.

In [248]:
ins_2016 = ins3[ins3['year'] == 2016]
ins_2016

business_id  score       date     type  year
0               10     82 2016-05-03  routine  2016
3               19     94 2016-05-13  routine  2016
6               24     98 2016-10-05  routine  2016
7               24     96 2016-03-11  routine  2016
10              31     84 2016-08-16  routine  2016
...            ...    ...        ...      ...   ...
15425        89515     98 2016-11-03  routine  2016
15426        89701    100 2016-12-06  routine  2016
15427        89790     96 2016-11-29  routine  2016
15428        89803    100 2016-12-15  routine  2016
15429        89900     93 2016-12-06  routine  2016

[5839 rows x 5 columns]

### Question 6

#### Question 6a

Merge the business and 2016 inspections data, keeping all businesses regardless of whether they show up in the inspections file. Show the first several rows of the resulting dataframe.

In [249]:
merged_df = pd.merge(vb, ins3, on='business_id', how='left')
merged_df.head()

business_id                   name                       address  \
0           10       TIRAMISU KITCHEN                 033 BELDEN PL   
1           10       TIRAMISU KITCHEN                 033 BELDEN PL   
2           10       TIRAMISU KITCHEN                 033 BELDEN PL   
3           19  NRGIZE LIFESTYLE CAFE  1200 VAN NESS AVE, 3RD FLOOR   
4           19  NRGIZE LIFESTYLE CAFE  1200 VAN NESS AVE, 3RD FLOOR   

            city state postal_code   latitude   longitude  phone_number  \
0  San Francisco    CA       94104  37.791116 -122.403816  +14154217044   
1  San Francisco    CA       94104  37.791116 -122.403816  +14154217044   
2  San Francisco    CA       94104  37.791116 -122.403816  +14154217044   
3  San Francisco    CA       94109  37.786848 -122.421547  +14157763262   
4  San Francisco    CA       94109  37.786848 -122.421547  +14157763262   

   score       date     type    year  
0   82.0 2016-05-03  routine  2016.0  
1   94.0 2014-07-29  routine  2014.0  
2   92.0 2014-01-14  routine  2014.0  
3   94.0 2016-05-13  routine  2016.0  
4   94.0 2014-11-10  routine  2014.0

#### Question 6b
Print the 20 lowest rated businesses names, their addresses, and their ratings.

In [250]:
lowest_rated = merged_df.dropna(subset=['score']).sort_values(by='score', ascending=True).head(20)  # filter out rows with NaN score, and then sort by score
lowest_rated_businesses = lowest_rated[['name', 'address', 'score']]
lowest_rated_businesses

name                     address  \
10400                        ABC BAKERY CAFE             650 JACKSON ST    
6288                         IMPERIAL PALACE          818 WASHINGTON ST    
9098          YUMMY DIM SUM & FAST FOOD, LLC            930 STOCKTON ST    
11391                GOLDEN RIVER RESTAURANT            5827 GEARY BLVD    
2426                        YEE'S RESTAURANT             1131 GRANT AVE    
2446                             LES JOULINS                44 ELLIS ST    
8484        OZONE THAI RESTAURANT AND LOUNGE                598 02ND ST    
11723                        MISS SAIGON LLC                100 06TH ST    
6633                     BRISTOL FARMS #2317           845 MARKET ST 010   
2715             BASIL THAI RESTAURANT & BAR             1175 FOLSOM ST    
5884                           BODEGA BISTRO              607 LARKIN ST    
14091                HARBOR VILLA RESTAURANT  5238 DIAMOND HEIGHTS BLVD    
13966                           CRAZY PEPPER          2257 SAN JOSE AVE    
7645                              FIRST CAKE            5901 GEARY BLVD    
15130                              POKI TIME            2101 LOMBARD ST    
13919  LE SOLEIL INTERNATIONAL HOLDINGS, INC          133 CLEMENT  ST.     
11334                          SA BEANG THAI          312 DIVISADERO ST    
3362                         GOURMET KITCHEN           1051 STOCKTON ST    
11348                             UNCLE CAFE              65 WAVERLY PL    
12616                         RAMEN YAMADAYA           1728 BUCHANAN ST    

       score  
10400   46.0  
6288    47.0  
9098    50.0  
11391   52.0  
2426    53.0  
2446    53.0  
8484    54.0  
11723   54.0  
6633    54.0  
2715    54.0  
5884    54.0  
14091   55.0  
13966   55.0  
7645    55.0  
15130   55.0  
13919   55.0  
11334   55.0  
3362    55.0  
11348   55.0  
12616   56.0

## Done!

Now commit this notebook to your `hw3_submissions` branch, push it to your GitHub repo, and open a PR!